### PR TITLE
7 implement ot

### PR DESCRIPTION
**Pull request by joannacknight**
The main features of the PR are:

- Introduced a new `DistanceMetric` class
- Removed the `mmd.py` script and created MMD as a child class of `DistanceMetric`
- Introduced a new `OTDD` class that calculates the otdd between two datasets
- Updated the `compute_similarity` method of `DMPair` so that it returns labels as well as the data, also it now returns all data as Numpy arrays (previously it could be either Numpy or a Tensor depending on whether it was only using train data or not)
- Introduced a `test_otdd.py `script for testing the `OTDD` class

Some thoughts on testing:
- The PyTest tests can be run as they are but further otdd configurations could be added into the `metrics.yaml` file - currently there are only two options there and there are a lot of parameters to experiment with
- As the mmd code has changed, it is necessary to verify this still works and returns the expected results
- When I was testing I experienced the otdd tests failing inconsistently for the `otdd_naive_upperbound` test parameters. After setting a random seed when initialising the `OTDD` object the tests no longer failed... however, I am not 100% sure if this is a definite fix.
- The ``calculate_metrics.py`` script also needs testing

Some further work that needs considering after this:
- Further safeguarding to prevent metrics being calculated on Apple M1 cpu (raised issue #45 )
- The parameters to be set for `OTDD` - will the exact calculation be too computationally intensive? (raised issue #48)
- Do we want any other implementations of OT? If so, I envisage them as being set up as their own class rather than an option in the `OTDD` class (raised issue #49)

### CODE REVIEW

File: src/modsim2/similarity/metrics.py
Comment (lannelin) [inline src/modsim2/similarity/metrics.py:1]: maybe worth splitting into submodules so we have src/modsim2/similarity/metrics/mmd.py etc? can always hoist them up with an import later in an `__init__.py`

File: src/modsim2/data/loader.py [hunk 1]
```diff
@@ -365,29 +361,40 @@ def _split_indices(
...
+
+            labels_A = train_labels_A
+            labels_B = train_labels_B
 
         # Loop over dict, compute metrics
         similarity_dict = {}
         for key, metric in self.metric_config.items():
-            similarity_dict[key] = METRIC_FN_DICT[metric[FUNCTION]](
-                data_A, data_B, **metric[ARGUMENTS]
+            obj_metric = METRIC_CLS_DICT[metric[CLASS]](metric_seed)
```
[the remark this answers is not in the document]
Comment (joannacknight) [reply src/modsim2/data/loader.py:395]: `obj_metric` renamed `metric_conf`
`metric_conf` renamed `metrics_config`
`CLASS` renamed `CLASS_KEY` (from another comment)

[the remark this answers is not in the document]
Comment (joannacknight) [reply src/modsim2/data/loader.py:395]: Added named argument `seed=`
Separated out `metric_conf` assignment to two lines of code as suggested

File: src/modsim2/similarity/constants.py [hunk 2]
```diff
@@ -1,5 +1,8 @@
-from modsim2.similarity.mmd import mmd
+from modsim2.similarity import metrics
 
-FUNCTION = "function"
 ARGUMENTS = "arguments"
-METRIC_FN_DICT = {"mmd": mmd}
+CLASS = "class"
```
[the remark this answers is not in the document]
Comment (joannacknight) [reply src/modsim2/similarity/constants.py:4]: Done

File: src/modsim2/similarity/metrics.py
Comment (joannacknight) [reply src/modsim2/similarity/metrics.py:1]: Ignore my question - found an example in another part of the code.

File: src/modsim2/similarity/metrics.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+class DistanceMetric(ABC):
+    def __init__(self, seed: int = 42):
+        torch.manual_seed(seed)
+
+    @abstractmethod
+    def _pre_process_data(self):
+        pass
+
+    @abstractmethod
+    def calculate_metric(
```
[the remark this answers is not in the document]
Comment (joannacknight) [reply src/modsim2/similarity/metrics.py:23]: Agreed and re-named

File: tests/test_mmd.py [hunk 4]
```diff
@@ -10,6 +10,7 @@
 METRICS_CONFIG_PATH = os.path.join(PROJECT_ROOT, "tests", "testconfig", "metrics.yaml")
 CONFIGS = load_configs(METRICS_CONFIG_PATH)
 METRIC_CONFIG = CONFIGS["metric_config"]
+METRIC_CONFIG = {k: v for k, v in METRIC_CONFIG.items() if v["class"] == "mmd"}
```
[the remark this answers is not in the document]
Comment (joannacknight) [reply tests/test_mmd.py:13]: Your understanding is correct. There are other options rather than filtering:

1. We could just have one `test_metrics.py` file which would run the tests for all metrics. However, I think having separate scripts for each metric is quite useful for debugging, e.g. future development could cause just _one_ metric to fail in which case it's useful to have a test file to run that single metric while debugging.
2. We could split the `metrics.yaml` file into one file for each metric, i.e. `metrics_mmd.yaml` and `metrics_otdd.yaml`. But I thought it might be useful to see all the test params in one file. 

I didn't much like either of the above options and will implement the setup method as you have suggested.

File: src/modsim2/data/loader.py [hunk 5]
```diff
@@ -365,29 +361,40 @@ def _split_indices(
             cifar=self.cifar,
         )
 
-    def compute_similarity(self, only_train: bool = False):
+    def compute_similarity(
+        self, only_train: bool = False, metric_seed: int = 42
```
Comment (philswatton) [inline src/modsim2/data/loader.py:365]: Set default argument for `metric_seed` to `self.seed`

File: src/modsim2/data/loader.py [hunk 6]
```diff
@@ -365,29 +361,40 @@ def _split_indices(
...
+    ) -> Dict:
         """
         compute similarity between data of A and B
         only_train removes the validation data from this comparison
         """
-
         # coerce data into single tensor (not subset)
         # TODO issue-15 want to get data post-transform
```
Comment (philswatton) [inline src/modsim2/data/loader.py:372]: We missed this in past PRs but issue #15 has been resolved for a while so this comment should be removed

File: src/modsim2/similarity/metrics.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+import torch
+from otdd.pytorch.distance import DatasetDistance
+from sklearn import metrics
+from torch.utils.data import TensorDataset
+
+from modsim2.similarity.embeddings import EMBEDDING_FN_DICT
+
+
+class DistanceMetric(ABC):
+    def __init__(self, seed: int = 42):
```
Comment (philswatton) [inline src/modsim2/similarity/metrics.py:15]: We want to make sure the DMPair object controls the seeding (so when the user sets up a dataset config, the seed is used across data creation, model training, and metrics calculation - we've stuck to a single seed for simplicity), so best to avoid having a default arg here to make sure we're never accidentally using this?

File: src/modsim2/similarity/metrics.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+        data_A: np.ndarray,
+        data_B: np.ndarray,
+        labels_A: np.ndarray,
+        labels_B: np.ndarray,
+        max_samples: int,
+        device: str,
+        **kwargs,
+    ) -> float:
+        """
+        Calculates the optimal transport distance between datasets
```
Comment (philswatton) [inline src/modsim2/similarity/metrics.py:144]: Very very minor comment, but for clarity should be `optimal transport dataset distance`

File: tests/test_otdd.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+            similarity_dict[k]
+            == METRIC_CONFIG[k]["expected_results"]["same_result_only_train"]
+        )
+
+
+# This test checks that the distance between a dataset and itself is the expect value
+# When exact calculations are used then this will be zero
+# Approximate methods may be non-zero and these are checked against the known value
+# for the seed
+def test_cifar_otdd_different():
```
[the remark this answers is not in the document]
Comment (philswatton) [reply tests/test_otdd.py:37]: I also get a failure on these tests, but get a different result again to James

```
FAILED tests/test_otdd.py::test_cifar_otdd_same - assert 256.1786193847656 == 256.1785888671875
FAILED tests/test_otdd.py::test_cifar_otdd_different - assert 98.62093353271484 == 98.62129211425781
```

**Review by philswatton — CHANGES_REQUESTED** [review]
Overall looks really good and a huge improvement on the previous approach to the metrics. I've suggested some fairly minor changes

**Comment by philswatton** [comment]
Also, agreed re OT requiring a separate implementation to OTDD

File: src/modsim2/data/loader.py [hunk 10]
```diff
@@ -365,29 +361,40 @@ def _split_indices(
...
             data_A = np.concatenate((train_data_A, val_data_A), axis=0)
             data_B = np.concatenate((train_data_B, val_data_B), axis=0)
+
+            labels_A = np.concatenate((train_labels_A, val_labels_A), axis=0)
+            labels_B = np.concatenate((train_labels_B, val_labels_B), axis=0)
         else:
-            data_A = train_data_A
-            data_B = train_data_B
+            data_A = np.array(train_data_A)
+            data_B = np.array(train_data_B)
```
[the remark this answers is not in the document]
Comment (joannacknight) [reply src/modsim2/data/loader.py:387]: Yes, agreed that would be neater.

Just to note: as a result of this change I have had to cast them back to tensors for some tests in test_data.py

File: tests/test_otdd.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+
+# This test checks that the distance between a dataset and itself is the expect value
+# When exact calculations are used then this will be zero
+# Approximate methods may be non-zero and these are checked against the known value
+# for the seed
+def test_cifar_otdd_same():
+    dmpair = DMPair(metric_config=METRIC_CONFIG, seed=42)
+    similarity_dict = dmpair.compute_similarity(only_train=False, metric_seed=42)
+    for k in similarity_dict:
```
[the remark this answers is not in the document]
Comment (joannacknight) [reply tests/test_otdd.py:23]: The suggestion looks neater, but not sure it is more efficient? You would still have to loop over k, i.e. 

```
for k in METRIC_CONFIG:
    expected_trainval = METRIC_CONFIG[k]["expected_results"]["same_result"]
    expected_trainonly = METRIC_CONFIG[k]["expected_results"]["same_result_only_train"]
    assert dmpair.compute_similarity(only_train=False, metric_seed=42)[k] == expected_trainval
    assert dmpair.compute_similarity(only_train=True, metric_seed=42)[k] == expected_trainonly
```

And then compute_similarity will be calculated for value of k. Could create one `dmpair` for each `k` so that it's not recalculating compute_similarity for all k each time, but that would mean the datasets are created for each `k` which isn't good either. 

I could reduce it to one `for` loop by naming the `similarity_dict` variables differently.

File: src/modsim2/data/loader.py (hunk 5 above)
Comment (joannacknight) [reply src/modsim2/data/loader.py:365]: @philswatton Should `metric_seed` be an argument in this method at all? e.g. do we ever want to set it independently of the dmpair seed? 

Also a note: it isn't possible to set the default directly to `self.seed` (get error name 'self' not defined), so would need to set `metric_seed: int = None` in the method definition and then assign `metric_seed` to `self.seed` if `metric_seed is None`.

File: tests/test_mmd.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (joannacknight) [reply tests/test_mmd.py]: As there is nothing to teardown, I think the fixture can `return` rather than `yield` the dictionary?

File: src/modsim2/data/loader.py [hunk 12]
```diff
@@ -398,11 +403,12 @@ def compute_similarity(self, only_train: bool = False):
     @staticmethod
     def _get_subset_data(
         subset_module: CIFAR10DMSubset,
-    ) -> Tuple[torch.Tensor, torch.Tensor]:
+    ) -> Tuple[np.array, np.array]:
         # force __getitem__ in CIFAR10 class for train and val
         # just take image rather than label
         train = torch.stack([item[0] for item in subset_module.dataset_train])
         val = torch.stack([item[0] for item in subset_module.dataset_val])
+        train, val = np.array(train), np.array(val)
```
Comment (philswatton) [inline src/modsim2/data/loader.py:411]: Instead of stacking via torch and then passing into `np.array`, we can just use `np.stack` in the first place:

```suggestion
        train = np.stack([item[0] for item in subset_module.dataset_train])
        val = np.stack([item[0] for item in subset_module.dataset_val])
```

Broadly, I think fine to have an output of either `torch.tensor` or `np.array` here, but from a clarity perspective it would be good to avoid changing it further in `compute_similarity`. Instead, we should have a default type that's clear in this fn, and in places where we need it (e.g. MMD needs `np.array` if that's not the default, OTDD needs `torch.tensor` if that's not the default) change it there.

File: tests/test_otdd.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    # filter down to only otdd configs
+    otdd_config = {k: v for k, v in otdd_config.items() if v["class"] == "otdd"}
+    return otdd_config
+
+
+# This test checks that the distance between a dataset and itself is the expect value
+# When exact calculations are used then this will be zero
+# Approximate methods may be non-zero and these are checked against the known value
+# for the seed
+def test_cifar_otdd_same(metrics_config: dict):
```
Comment (philswatton) [inline tests/test_otdd.py:26]: I'm still getting a failure on both tests even with a fresh installation of everything, so we may need to settle for the 'within a range' test for the OTDD tests!

File: src/modsim2/similarity/metrics/mmd.py [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+logger = logging.getLogger(__name__)
+
+
+class MMD(DistanceMetric):
+    def __init__(self, seed: int):
+        # Kernel dictionary
+        self.__MMD_KERNEL_DICT = {
+            "rbf": metrics.pairwise.rbf_kernel,
+            "laplace": metrics.pairwise.laplacian_kernel,
+        }
```
Comment (philswatton) [inline src/modsim2/similarity/metrics/mmd.py:20]: Missing `super().__init__(seed)` call

**Review by philswatton — CHANGES_REQUESTED** [review]
Looks good. The merge needs to be done and resolved, and I have some additional comments (one of which I really should have thought of last time!). Overall looks better again!

File: src/modsim2/data/loader.py (hunk 5 above)
Comment (joannacknight) [reply src/modsim2/data/loader.py]: Think I have answered my question on this one - as it is possible to call compute_similarity multiple times for one DMPair, then it makes sense to be able to set the `metric_seed` independently of the object's seed.

File: tests/test_otdd.py (hunk 13 above)
Comment (joannacknight) [reply tests/test_otdd.py:32]: As discussed, the next update will change the tests to check for 'closeness' rather than equality

File: src/modsim2/similarity/metrics/mmd.py (hunk 14 above)
Comment (joannacknight) [reply src/modsim2/similarity/metrics/mmd.py:20]: I've added this in though thought the mmd calculations were deterministic

**Comment by joannacknight** [comment]
The updates to the otdd tests are now ready to be tested. 

I wanted to write the tests so that only one assert statement would be made for each test - it would be best practice to implement them that way so that all the metric configs get tested. Currently if the first one fails, then the others don't run and I thought it would be useful when debugging to know if it's only one that has failed or all of them, for example. 

I thought this should be possible by parametrising the tests (https://docs.pytest.org/en/7.1.x/example/parametrize.html). I could write the tests in pseudo-code and then had an L&D day on Friday trying to get this working in Pytest, however it doesn’t seem possible to use fixtures directly as arguments, e.g. I wanted to write something like the following, but it throws an error due to metrics_config being a fixture:
`@pytest.mark.parametrize(‘metric_config’, [metric_config for metric_config in metrics_config])`

I couldn’t find a way around it so I’ve implemented a work-around that still runs multiple checks in each test, but it will run all checks and then print out the details of any failures. You need to run pytest in verbose mode (`-v`) in order to see the details of the failures if there are many. My preference is to run `pytest --disable-warnings -v tests` to get rid of all the warning messages too.

I don't mind changing the code back to the multiple assert statements in one test if the preference is to keep them consistent with the mmd tests. 

A few notes:
- I have set the expected values for the otdd tests in the `metrics.yaml` file to be the values that James posted in slack
- With the expected values set as they are:
  - Two otdd tests should be skipped for Phil, and the third (checking that an error is thrown) should pass
  - Two otdd tests should pass for James, and the third should be skipped
- I have used the default tolerance values in the `isclose` calculation - do these need changing? May depend on James’ results
- It is obviously possible to amend the expected values in the `metrics.yaml` file and the code in the `skip` statements to test the tests using different scenarios.

File: src/modsim2/similarity/metrics/mmd.py (hunk 14 above)
Comment (philswatton) [reply src/modsim2/similarity/metrics/mmd.py]: Oh true! My bad!

**Comment by lannelin** [comment]
tests are failing for me.
Macbook:
```
E           Failed: test:otdd_exact/same_result, expected result: 0.0, actual result: 98.62165069580078
E           test:otdd_naive_upperbound/same_result, expected result: 256.1785888671875, actual result: 306.5455017089844
E           test:otdd_exact/same_result_only_train, expected result: 0.0, actual result: 98.76387786865234
E           test:otdd_naive_upperbound/same_result_only_train, expected result: 256.1785888671875, actual result: 306.96649169921875
```

azure:
```
E           Failed: test:otdd_exact/same_result, expected result: 0.0, actual result: 98.62165069580078
E           test:otdd_naive_upperbound/same_result, expected result: 256.1785888671875, actual result: 306.5455017089844
E           test:otdd_exact/same_result_only_train, expected result: 0.0, actual result: 98.76387023925781
E           test:otdd_naive_upperbound/same_result_only_train, expected result: 256.1785888671875, actual result: 306.96649169921875
```

slight difference there in otdd_exact/same_result_only_train!

aside, is the expected val for `same_result_only_train` worryingly high, perhaps not?

File: tests/test_otdd.py [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+)
+def test_cifar_otdd_different(metrics_config: dict):
+    dmpair = DMPair(metrics_config=metrics_config, drop_percent_A=0.2, seed=42)
+    similarity_dict = dmpair.compute_similarity(only_train=False)
+    similarity_dict_only_train = dmpair.compute_similarity(only_train=True)
+    test_scenarios = {
+        "diff_result": similarity_dict,
+        "diff_result_only_train": similarity_dict_only_train,
+    }
+    failures = []
```
Comment (lannelin) [inline tests/test_otdd.py:81]: I like the reporting that we get out of this approach.
You might want to consider using something like `pytest-check` that achieves a similar outcome. 
```suggestion
    from pytest_check import check # move this up
    
    for scenario, results in test_scenarios.items():
        for k in metrics_config:
            expected_result = metrics_config[k]["expected_results"][scenario]
            actual_result = results[k]

            with check:
                assert np.isclose(actual_result, expected_result, rtol=1e-5, atol=1e-8)
```

if you run pytest in verbose then you still get the detailed output of mismatched values as you've captured here.

If y

Comment (lannelin) [reply tests/test_otdd.py:81]: happy to keep this if you'd prefer not to introduce another dependency.
Whichever approach is taken, we should do the same in the mmd tests.

**Review by lannelin — COMMENTED** [review]
looks great and passes on my Mac and Azure.

Have made a suggestion on the tests structure. Happy with whatever you choose but should be consistent across metrics.

Nice work.

**Review by lannelin — APPROVED** [review]
